src/valuetype.jl

```diff
@@ -0,0 +1,71 @@
+
+
+"""
+    valuetype(x)
+
+Return the type of the underlying value contained in `x`.
+
+For primitive types like `Number`, this is the type of `x` itself.
+For container types like `AbstractArray{T}`, this is `T`.
+"""
+function valuetype end
+valuetype(::Type{<:Quantity{T}}) where T  = T
+valuetype(::Type{T}) where T <: Real = T
+valuetype(::T) where T = valuetype(T)
+valuetype(T::DataType) = error("The valuetype function is not implemented for type $T")
+valuetype(::Type{<:AbstractArray{T}}) where T = T
+
+"""
+    change_valuetype(::Type{T}, x)
+
+Change the type of the value contained in `x` to `T`.
+
+If `x` is a primitive type like `Number`, this will convert `x` to `T` and return it.
+If `x` is a container type like `SVector{T}`, this will convert the elements of `x` to `T`.
+"""
+function change_valuetype end
+change_valuetype(::Type{T}, x) where {T} = convert(T, x)::T
+change_valuetype(::Type{T}, x::SVector{N}) where {T, N} = convert(SVector{3, T}, x)
+
+"""
+    common_valuetype(::Type{BaseType}, ::Type{DefaultType}, args...)
+
+Determine the common value type of the arguments `args...`, ensuring it is a subtype of `BaseType`.
+For args that are containers, such as `AbstractArray{T}`, the common value type is determined by the element type `T`.
+
+If the promoted value type of `args...` is not a subtype of `BaseType`, then `DefaultType` is returned instead.
+
+# Arguments
+- `BaseType::Type`: The required base type for the common value type.
+- `DefaultType::Type`: The fallback type if the common value type is not a subtype of `BaseType`.
+- `args...`: Arguments from which to determine the common value type.
+
+# Returns
+The common value type of `args...`, or `DefaultType` if the common type is not a subtype of `BaseType`.
+"""
+function common_valuetype end
+function common_valuetype(::Type{BaseType}, ::Type{DefaultType}, args::Vararg{Any, N}) where {BaseType, DefaultType, N}
+    T = promote_type(valuetype.(args)...)
```

Comment:
In order to be consistent with the last function using map, and also a bit because of what we discussed where broadcast might infer worse than map (see e.g. here: https://discourse.julialang.org/t/map-vs-broadcast-should-one-prefer-map-if-these-are-equivalent/86118/2).

I'd change this to `T = promote_type(map(valuteyp, args)...)`.

Anything against this?